src/urllib3/multipart/decoder.py

```diff
@@ -0,0 +1,165 @@
+"""Logic for parsing and decomposing a multipart response body."""
+import email.parser
+import typing
+
+from .. import _collections
+from .encoder import encode_with
+
+if typing.TYPE_CHECKING:
+    from .. import response as _response
+
+
+@typing.overload
+def _split_on_find(content: str, bound: str) -> typing.Tuple[str, str]:
+    ...
+
+
+@typing.overload
+def _split_on_find(content: bytes, bound: bytes) -> typing.Tuple[bytes, bytes]:
+    ...
+
+
+def _split_on_find(
+    content: typing.AnyStr, bound: typing.AnyStr
+) -> typing.Tuple[typing.AnyStr, typing.AnyStr]:
+    point = content.find(bound)
+    return content[:point], content[point + len(bound) :]
+
+
+class ImproperBodyPartContentError(Exception):
+    pass
+
+
+class NonMultipartContentTypeError(Exception):
+    pass
+
+
+def _header_parser(
+    headers: bytes, encoding: str
+) -> typing.Sequence[typing.Tuple[str, str]]:
+    string = headers.decode(encoding)
+    items = email.parser.HeaderParser().parsestr(string).items()
+    items = typing.cast(typing.List[typing.Tuple[str, str]], items)
```

Comment:
If it's only changing to match the return value you can use a type ignore to avoid a call to typing.cast()

Reply:
Yeah, apparently `HeaderParser().parsestr()` returns a `Mapping[str, Any]` but I don't know why it uses that :shrug:

Generally, I find `cast` to be the explicit and `type: ignore` to be the implicit and the former explains why we're doing it, the latter looks like we're satisfying a computer gone wrong

Reply:
Looks like `email.parser.HeaderParser().parsestr()` returns `email.message.Message` which looks like it's trying to be a `Mapping`.

[From typeshed](https://github.com/python/typeshed/blob/master/stdlib/email/message.pyi) it looks like `email.message.Message.items()` returns `List[Tuple[str, Any]]`, which explains why we need the cast. Carry on!